### PR TITLE
ff: groebner bases for incremental ideals

### PR DESCRIPTION
**Pull request by alex-ozdemir**

### CODE REVIEW

File: src/theory/ff/groebner.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  const std::vector<CoCoA::RingElem>& solution();
+  // Remove the last batch of generators
+  void pop();
+
+ private:
+  void ensureSolution();
+
+  std::unique_ptr<context::Context> d_context;
+
+  CoCoA::ring d_polyRing;
```
Comment (ajreynol) [inline src/theory/ff/groebner.h:66]: Can you give some quick doc on these members?

What context do they depend on? When are they computed? Etc.

File: src/theory/ff/groebner.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  std::vector<CoCoA::RingElem> trivialCore();
+  // For a non-trivial idea, check whether there is a base-field variety member.
+  bool hasSolution();
+  // For a non-trivial idea with a base-field variety member, get it.
+  const std::vector<CoCoA::RingElem>& solution();
+  // Remove the last batch of generators
+  void pop();
+
+ private:
+  void ensureSolution();
```
Comment (ajreynol) [inline src/theory/ff/groebner.h:62]: Doc.

File: src/theory/ff/groebner.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+  std::unique_ptr<context::Context> d_context;
```
Comment (ajreynol) [inline src/theory/ff/groebner.h:64]: What is this used for?

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: test/unit/theory/theory_ff_groebner_black.cpp [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+  ideal.pushGenerators({a * (a - 1) - b});
+  EXPECT_EQ(ideal.idealIsTrivial(), false);
+  EXPECT_EQ(ideal.hasSolution(), true);
+  ideal.pushGenerators({b * (a - 2) - 1});
+  EXPECT_EQ(ideal.idealIsTrivial(), false);
+  EXPECT_EQ(ideal.hasSolution(), false);
+}
+
+#endif  // CVC5_USE_COCOA
```
Comment (ajreynol) [inline test/unit/theory/theory_ff_groebner_black.cpp:201]: This has the same issue with the guard :)

File: src/theory/ff/groebner.h (hunk 3 above)
Comment (alex-ozdemir) [reply src/theory/ff/groebner.h:85]: Good question. Documented.

File: test/unit/theory/theory_ff_groebner_black.cpp (hunk 4 above)
Comment (alex-ozdemir) [reply test/unit/theory/theory_ff_groebner_black.cpp]: Good catch :facepalm:

**Comment by alex-ozdemir** [comment]
Thanks for the comment, Andy. Indeed the documentation was not up to scruff. Let me know what you think now, and whether you have more comments.

File: test/unit/theory/theory_ff_groebner_black.cpp [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+#ifdef CVC5_USE_COCOA
+#include <CoCoA/BigInt.H>
+#include <CoCoA/QuotientRing.H>
+#include <CoCoA/RingZZ.H>
+#include <CoCoA/SparsePolyOps-ideal.H>
+#include <CoCoA/SparsePolyRing.H>
+#include <CoCoA/ring.H>
+#include <CoCoA/symbol.H>
+#endif  // CVC5_USE_COCOA
```
Comment (ajreynol) [inline test/unit/theory/theory_ff_groebner_black.cpp:24]: Like the other unit tests, let's just guard the whole file.

**Review by ajreynol — CHANGES_REQUESTED** [review]
One minor comment, else LGTM.

**Review by ajreynol — APPROVED** [review]

File: test/unit/theory/theory_ff_groebner_black.cpp (hunk 5 above)
Comment (alex-ozdemir) [reply test/unit/theory/theory_ff_groebner_black.cpp]: Whoops. Good catch.

**Pull request merged**